File: engine/src/Strategy.ts
// 玩家策略接口 + 常用实现
// 策略在 BattleSim 每 tick 被调用一次，返回想要建造的 building id
// 无法建造时返回 null (攒钱或等槽位)

import type { Building } from "./Building.ts";

export interface DecisionContext {
  now: number;
  gold: number;
  income: number;
  availableSlots: number;
  ownedBuildings: readonly Building[];
  /** 查建筑造价；找不到 building 或不够钱时返回 false */
  canAfford: (buildingId: string) => boolean;
  /** 已拥有某 building id 的数量 */
  ownedCountOf: (buildingId: string) => number;
}

export interface PlayerStrategy {
  name: string;
  decide(ctx: DecisionContext): string | null;
}

/**
 * 固定出牌顺序：按列表依次建造。loop=true 时建完一轮重新开始 (继续刷同样的兵营)
 */
export function buildOrder(name: string, order: readonly string[], loop: boolean = true): PlayerStrategy {
  let idx = 0;
  return {
    name,
    decide(ctx) {
      if (order.length === 0) return null;
      if (idx >= order.length) {
        if (!loop) return null;
        idx = 0;
      }
      const wanted = order[idx]!;
      if (!ctx.canAfford(wanted)) return null;       // 没钱 → 等
      if (ctx.availableSlots <= 0) return null;      // 没槽 → 等
      idx++;
      return wanted;
    },
  };
}

/**
 * 贪心：维护一个目标配比 (building id → 想要多少个)，每次挑"缺口最大且买得起"的建
 */
export function mixedComposition(name: string, target: Record<string, number>): PlayerStrategy {
  return {
    name,
    decide(ctx) {
      if (ctx.availableSlots <= 0) return null;
      let bestId: string | null = null;
      let bestGap = 0;
      for (const id of Object.keys(target)) {
        const gap = (target[id] ?? 0) - ctx.ownedCountOf(id);
        if (gap > bestGap && ctx.canAfford(id)) {
          bestGap = gap;
          bestId = id;
        }
      }
      return bestId;
    },
  };
}

/** 什么都不造 (用于测试 AI 被推塔) */
export const doNothing: PlayerStrategy = { name: "idle", decide: () => null };

File: unity/Assets/Mg01/Sim/Strategy.cs
// 与 engine/src/Strategy.ts 一一对应
// BuildOrder + MixedComposition + DoNothing 三个内置策略

using System.Collections.Generic;

namespace Mg01.Sim
{
    /// <summary>固定出牌：按列表依次造；loop=true 时建完一轮重头开始</summary>
    public class BuildOrderStrategy : PlayerStrategy
    {
        public string name;
        private readonly string[] _order;
        private readonly bool _loop;
        private int _idx;

        public BuildOrderStrategy(string name, string[] order, bool loop = true)
        {
            this.name = name;
            this._order = order;
            this._loop = loop;
        }

        public string Decide(StrategyContext ctx)
        {
            if (_order.Length == 0) return null;
            if (_idx >= _order.Length)
            {
                if (!_loop) return null;
                _idx = 0;
            }
            string wanted = _order[_idx];
            if (!ctx.canAfford(wanted)) return null;
            if (ctx.availableSlots <= 0) return null;
            _idx++;
            return wanted;
        }
    }

    /// <summary>贪心：维护目标配比 (id → 想要数量)，每次挑缺口最大且买得起的造</summary>
    public class MixedCompositionStrategy : PlayerStrategy
    {
        public string name;
        private readonly Dictionary<string, int> _target;

        public MixedCompositionStrategy(string name, Dictionary<string, int> target)
        {
            this.name = name;
            this._target = target;
        }

        public string Decide(StrategyContext ctx)
        {
            if (ctx.availableSlots <= 0) return null;
            string bestId = null;
            int bestGap = 0;
            foreach (var kv in _target)
            {
                int gap = kv.Value - ctx.ownedCountOf(kv.Key);
                if (gap > bestGap && ctx.canAfford(kv.Key))
                {
                    bestGap = gap;
                    bestId = kv.Key;
                }
            }
            return bestId;
        }
    }

    /// <summary>什么都不造（用于测试 AI 被推塔）</summary>
    public class DoNothingStrategy : PlayerStrategy
    {
        public string Decide(StrategyContext ctx) => null;
    }
}
